exchanges/orderbook/node.go

```diff
@@ -0,0 +1,111 @@
+package orderbook
+
+import (
+	"sync/atomic"
+	"time"
+)
+
+const (
+	cleanerActive    = 1
+	cleanerNotActive = 0
+)
+
+var (
+	defaultInterval  = time.Minute
+	defaultAllowance = time.Second * 30
+)
+
+// node defines a linked list node for an orderbook item
+type node struct {
+	value Item
+	next  *node
+	prev  *node
+	// Denotes time pushed to stack, this will influence cleanup routine when
+	// there is a pause or minimal actions during period
+	shelved time.Time
+}
+
+// stack defines a FIFO list of reusable nodes
+type stack struct {
+	nodes []*node
+	sema  uint32
+	count int32
+}
+
+// newStack returns a ptr to a new stack instance, also starts the cleaning
+// service
+func newStack() *stack {
+	s := &stack{}
+	go s.cleaner()
+	return s
+}
+
+// Push pushes a node pointer into the stack to be reused the time is passed in
+// to allow for inlining which sets the time at which the node is theoretically
+// pushed to a stack.
+func (s *stack) Push(n *node, tn time.Time) {
```

Comment:
time.Now() doesn't inline so I would rather inline the push function and do one time.Now() function call, instead of function call push and then function call time.Now(). If they are nanoseconds apart or microseconds apart, it doesn't really make a big difference as the cleanup is 30-second intervals anyway, if it's sitting on the stack for that long it's going to be cleaned up eventually. I see your point though but in this instance, I don't see an actual gain for being precise.

Reply:
Yeah in the end, since its inlined, the time.Now() delay isnt going to be much. The main thing is still around taking a variable time.Time, opening the possibility of variance. Its not happening now, but the only value you want is `time.Now()` but this opens up the possibility of other values being passed in which gives potential longer term debug fun. I understand you are keen on inlining these functions, so would you be more open to handling the `n`'s properties before pushing and removing the time variable?
This still allows for inlining
```
func (n *node) prepareForShelving() {
	n.next = nil
	n.prev = nil
	n.value = Item{}
	n.shelved = time.Now()
}
```
and then its 
```
n.prepareForShelving()
stack.Push(n)
```
Or even something that returns `n` so that you can call `stack.Push(prepareForShelving(n))`

Reply:
Added type for time and constructor so as to not change much.